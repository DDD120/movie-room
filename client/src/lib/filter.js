import dayjs from "dayjs";

export const arrayDeduplication = (list) => {
  return list.filter(
    (item, index, array) =>
      index === array.findIndex((data) => data.id === item.id)
  );
};

export const getYear = (date) => {
  return dayjs(date).year();
};
