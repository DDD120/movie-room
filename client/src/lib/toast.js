import { toast } from "react-toastify";

export const showToast = (message) => {
  toast(message, {
    autoClose: 3000,
    hideProgressBar: true,
  });
};
