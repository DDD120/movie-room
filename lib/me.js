export const me = (user) => {
  return {
    id: user._id,
    profile: {
      nickname: user.profile.nickname,
      thumbnail: user.profile.thumbnail,
    },
    email: user.email,
    reviews: user.reviews,
  };
};
