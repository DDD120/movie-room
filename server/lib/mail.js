import nodemailer from "nodemailer";
import dotenv from "dotenv";

dotenv.config();

export const mailSender = async (email, certificationNumber) => {
  let transporter = nodemailer.createTransport({
    service: "gmail",
    host: "smtp.gmail.com",
    port: 587,
    secure: false,
    auth: {
      user: process.env.NODEMAILER_USER,
      pass: process.env.NODEMAILER_PASS,
    },
  });

  try {
    let info = await transporter.sendMail({
      from: `"MOVIE ROOM" <${process.env.NODEMAILER_USER}>`,
      to: email,
      subject: "[MOVIE ROOM] 회원가입 인증번호 발송",
      html: `
        <body>
          <h1>[MOVIE ROOM]</h1>
          </br>
          <p>MOVIE ROOM에서 인증번호를 발송해 드립니다.</p>
          <div style="padding: 12px 0;">
            <span style="font-size: 24px; font-weight; bold;">${certificationNumber}</span>
          </div>
        </body>
      `,
    });

    return info.accepted;
  } catch (e) {
    throw new Error("메일 전송에 실패하였습니다.", console.error(e));
  }
};
