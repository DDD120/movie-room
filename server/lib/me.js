export const me = (user) => {
  return {
    id: user._id,
    nickname: user.nickname,
    thumbnail: user.thumbnail,
    email: user.email,
  };
};
