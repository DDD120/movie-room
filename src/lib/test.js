import React from "react";
import { render } from "@testing-library/react";
import { configureStore } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import userReducer from "store/user";
import { BrowserRouter } from "react-router-dom";
import { serverApi } from "apis/server-api";
import { moviedbApi } from "apis/movie-db-api";

export function renderWithProviders(
  ui,
  {
    preloadedState = {},
    store = configureStore({
      reducer: {
        user: userReducer,
        [serverApi.reducerPath]: serverApi.reducer,
        [moviedbApi.reducerPath]: moviedbApi.reducer,
      },
      preloadedState,
    }),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return (
      <Provider store={store}>
        <BrowserRouter>{children}</BrowserRouter>
      </Provider>
    );
  }

  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
