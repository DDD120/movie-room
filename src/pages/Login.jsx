import styled from "@emotion/styled";
import Container from "components/common/Container";
import { Common } from "styles/common";
import { Link, useNavigate } from "react-router-dom";
import { useRef } from "react";
import { useLoginMutation } from "apis/server-api";
import { useEffect, useCallback } from "react";
import { toast } from "react-toastify";
import { useDispatch } from "react-redux";
import { setUser } from "store/user";

const Layout = styled.div`
  display: flex;
  justify-content: center;
  flex-direction: column;
  align-items: center;
`;

const Logo = styled.div`
  width: 200px;
  margin: 40px 0 10px;
  & > img {
    width: 100%;
  }
`;

const Head = styled.h1`
  font-weight: 700;
  font-size: 1.5rem;
  color: ${Common.colors.black};
  margin: 10px 0;
`;

const Form = styled.form`
  width: 100%;
  display: flex;
  flex-direction: column;
  align-items: center;
`;

const Input = styled.input`
  margin: 4px auto;
  max-width: 600px;
  width: 100%;
  border-radius: 50px;
  padding: 14px 30px;
  border: none;
  box-shadow: none;
  color: #fff;
  font-size: 1.025rem;
  cursor: ${({ submit }) => submit && "pointer"};
  background-color: ${({ submit }) =>
    submit ? `${Common.colors.cyan}` : `${Common.colors.orange}`};
  &::placeholder {
    color: #fff;
  }
  &:-webkit-autofill,
  :-webkit-autofill:hover,
  :-webkit-autofill:focus,
  :-webkit-autofill:active {
    box-shadow: 0 0 0 1000px ${Common.colors.orange} inset;
    transition: background-color 5000s ease-in-out 0s;
    -webkit-box-shadow: 0 0 0 1000px ${Common.colors.orange} inset;
    -webkit-text-fill-color: #fff;
    -webkit-transition: background-color 5000s ease-in-out 0s;
  }
`;

const Signin = styled.span`
  display: block;
  margin-bottom: 10px;
`;

const Social = styled.a`
  position: relative;
  margin: 4px 0;
  max-width: 600px;
  height: 50px;
  width: 100%;
  border-radius: 12px;
  border: none;
  box-shadow: none;
  color: ${({ color }) => color};
  background-color: ${({ bgColor }) => bgColor};
  overflow: hidden;
  display: flex;
  justify-content: center;
  align-items: center;
  & > img {
    position: absolute;
    top: 0;
    left: 0;
    max-width: 50px;
  }
`;

const Login = () => {
  const emailRef = useRef();
  const passwordRef = useRef();

  const dispatch = useDispatch();
  const navigate = useNavigate();

  const [login, { data: loginRes = {}, isSuccess, error }] = useLoginMutation();

  const handleLoginSubmit = async (e) => {
    e.preventDefault();
    const email = emailRef.current.value;
    const password = passwordRef.current.value;
    if (email && password) {
      await login({ email, password });
    }
  };

  const showSuccessNotify = useCallback((msg) => {
    toast.success(msg, {
      autoClose: 3000,
      hideProgressBar: true,
    });
  }, []);

  const showErrorNotify = useCallback((msg) => {
    toast.error(msg, {
      autoClose: 3000,
      hideProgressBar: true,
    });
  }, []);

  useEffect(() => {
    if (loginRes.type === "SUCCESS_LOGIN") {
      console.log(loginRes);
      navigate("/");
      dispatch(setUser({ user: loginRes.user }));
      showSuccessNotify(loginRes.msg);
    }
    error && showErrorNotify(loginRes.msg);
  }, [
    error,
    isSuccess,
    navigate,
    loginRes,
    showSuccessNotify,
    showErrorNotify,
    dispatch,
  ]);

  return (
    <Container>
      <Layout>
        <Logo>
          <img src="/assets/logo.png" alt="로고" />
        </Logo>
        <Head>로그인</Head>
        <Form onSubmit={handleLoginSubmit} action="">
          <Input
            ref={emailRef}
            type="email"
            title="이메일"
            placeholder="이메일"
            autoFocus
          />
          <Input
            ref={passwordRef}
            type="password"
            title="비밀번호"
            placeholder="비밀번호"
          />
          <Input submit type="submit" title="로그인" value="로그인" />
        </Form>
        <Link to="/signup">
          <Signin>회원가입</Signin>
        </Link>
        <Social color={"#fff"} bgColor={"#03C75A"} href="">
          <img src="/assets/naver-symbol.png" alt="네이버 심볼" />
          <div>네이버 로그인</div>
        </Social>
        <Social color={"#000D9"} bgColor={"#FEE500"} href="">
          <img src="/assets/kakao-symbol.png" alt="카카오 심볼" />
          <div>카카오 로그인</div>
        </Social>
      </Layout>
    </Container>
  );
};

export default Login;
