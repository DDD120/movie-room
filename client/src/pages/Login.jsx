import styled from "@emotion/styled";
import Container from "components/common/Container";
import { colors, fontSize } from "styles/common";
import { Link, useSearchParams } from "react-router-dom";
import { useLoginMutation } from "apis/server-api";
import { useEffect } from "react";
import { useDispatch } from "react-redux";
import { login as storeLogin } from "store/user";
import { useForm } from "react-hook-form";
import { showToast } from "lib/toast";
import AuthInput from "components/common/AuthInput";
import LogoImg from "assets/logo.png";
import useAuthRedirect from "hooks/useAuthRedirect";

const Base = styled.div`
  display: flex;
  justify-content: center;
  flex-direction: column;
  align-items: center;
`;

const Logo = styled.div`
  width: 200px;
  margin: 40px 0 10px;
  img {
    width: 100%;
  }
`;

const Head = styled.h1`
  font-weight: 700;
  font-size: ${fontSize.lg};
  color: ${colors.black};
  margin: 10px 0;
`;

const Form = styled.form`
  width: 100%;
  display: flex;
  flex-direction: column;
  align-items: center;
`;

const Signin = styled.span`
  display: block;
  margin-bottom: 10px;
`;

const Login = () => {
  const {
    register,
    handleSubmit,
    formState: { isSubmitting },
  } = useForm();
  const [login, { data: loginRes = {}, isSuccess, isError, error }] =
    useLoginMutation();

  const dispatch = useDispatch();
  const [searchParams] = useSearchParams();
  const next = searchParams.get("next") ?? "/";
  useAuthRedirect();

  const toSignupPath = `/signup${next === "/" ? "" : `?next=${next}`}`;

  const handleLoginSubmit = ({ email, password }) => {
    login({ email, password });
  };

  useEffect(() => {
    if (isSuccess) {
      dispatch(storeLogin({ user: loginRes.user }));
      showToast(loginRes?.message);
    }
    if (isError) showToast(error.data?.message);
  }, [loginRes, isSuccess, dispatch, isError, error]);

  return (
    <Container>
      <Base>
        <Logo>
          <img src={LogoImg} alt="MOVIE ROOM 로고" />
        </Logo>
        <Head>로그인</Head>
        <Form onSubmit={handleSubmit(handleLoginSubmit)}>
          <AuthInput
            register={register("email", { required: true })}
            type="email"
            title="이메일"
            placeholder="이메일"
            autoFocus
          />
          <AuthInput
            register={register("password", { required: true })}
            type="password"
            title="비밀번호"
            placeholder="비밀번호"
          />
          <AuthInput
            submit
            type="submit"
            title="로그인"
            value="로그인"
            disabled={isSubmitting}
          />
        </Form>
        <Link to={toSignupPath}>
          <Signin>회원가입</Signin>
        </Link>
      </Base>
    </Container>
  );
};

export default Login;
