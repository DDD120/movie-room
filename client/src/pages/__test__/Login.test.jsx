import { fireEvent, screen } from "@testing-library/react";
import { renderWithProviders } from "lib/test";
import Login from "pages/Login";

describe("Login Page", () => {
  const setup = () => {
    return renderWithProviders(<Login />);
  };

  it("renders correctly", () => {
    const { container } = setup();
    expect(container).toMatchSnapshot();
  });

  it("Input value output test", async () => {
    setup();

    const emailValue = "email1234@mock.com";
    const passwordValue = "password1234";

    const email = screen.getByPlaceholderText("이메일");
    const password = screen.getByPlaceholderText("비밀번호");

    fireEvent.change(email, { target: { value: "email1234@mock.com" } });
    fireEvent.change(password, { target: { value: "password1234" } });

    expect(emailValue).toBe("email1234@mock.com");
    expect(passwordValue).toBe("password1234");
  });
});
