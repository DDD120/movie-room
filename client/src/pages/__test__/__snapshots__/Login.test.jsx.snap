// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Page renders correctly 1`] = `
<div>
  <div
    class="css-1ussbir"
  >
    <div
      class="css-jxj5pd"
    >
      <div
        class="css-o5wame"
      >
        <img
          alt="MOVIE ROOM 로고"
          src="/assets/logo.png"
        />
      </div>
      <h1
        class="css-wk5jya"
      >
        로그인
      </h1>
      <form
        class="css-13stbb8"
      >
        <input
          class="css-1a6crj1"
          name="email"
          placeholder="이메일"
          title="이메일"
          type="email"
        />
        <input
          class="css-1a6crj1"
          name="password"
          placeholder="비밀번호"
          title="비밀번호"
          type="password"
        />
        <input
          class="css-ol4lpg"
          title="로그인"
          type="submit"
          value="로그인"
        />
      </form>
      <a
        href="/signup"
      >
        <span
          class="css-hfj296"
        >
          회원가입
        </span>
      </a>
      <a
        class="css-11r631h"
        color="#fff"
        href=""
      >
        <img
          alt="네이버 심볼"
          src="/assets/naver-symbol.png"
        />
        <div>
          네이버 로그인
        </div>
      </a>
      <a
        class="css-1p3p2nj"
        color="#000D9"
        href=""
      >
        <img
          alt="카카오 심볼"
          src="/assets/kakao-symbol.png"
        />
        <div>
          카카오 로그인
        </div>
      </a>
    </div>
  </div>
</div>
`;
